frontend/catalyst/from_plxpr.py

```diff
@@ -0,0 +1,371 @@
+
+    See the documentation on
+    `Writing custom interpreters in JAX <https://jax.readthedocs.io/en/latest/notebooks/Writing_custom_interpreters_in_Jax.html>`_
+    for a walkthrough on the general architecture and behavior of this function.
+
+    Given that ``catalyst.jax_primitives.func_p`` does not define a concrete implementation, this
+    function will fail outside of an abstract evaluation call.
+
+    """
+    env = {}  # dict mapping var "variables" to val "values"
+
+    # Bind args and consts to environment
+    for arg, invar in zip(args, jaxpr.invars):
+        env[invar] = arg
+    for const, constvar in zip(consts, jaxpr.constvars):
+        env[constvar] = const
+
+    # Loop through equations and evaluate primitives using `bind`
+    for eqn in jaxpr.eqns:
+        # Read inputs to equation from environment
+        invals = [_read(invar, env) for invar in eqn.invars]
+        if eqn.primitive == qnode_prim:
+            if eqn.params["device"].shots != eqn.params["shots"]:
+                raise NotImplementedError("catalyst does not yet support dynamic shots")
+
+            f = partial(
+                _bind_catalxpr,
+                eqn.params["qfunc_jaxpr"].jaxpr,
+                eqn.params["qfunc_jaxpr"].consts,
+                eqn.params["device"],
+            )
+            # func_p is a CallPrimitive, so interpreter passed as first arg
+            # wrap_init turns the function into a WrappedFun, which can store
+            # transformations
+            outvals = func_p.bind(wrap_init(f), *invals, fn=eqn.params["qnode"])
+        else:
+            outvals = eqn.primitive.bind(*invals, **eqn.params)
+        # Primitives may return multiple outputs or not
+        if not eqn.primitive.multiple_results:
+            outvals = [outvals]
+        # Write the results of the primitive into the environment
+        for outvar, outval in zip(eqn.outvars, outvals):
+            env[outvar] = outval
+    return [env[outvar] for outvar in jaxpr.outvars]
+
+
+@dataclass
+class _InterpreterState:
+    """This dataclass stores the mutable variables modified
+    over the course of interpreting the plxpr as catalxpr."""
+
+    qreg: AbstractQreg
+    """The current quantum register."""
+
+    env: dict = field(default_factory=dict)
+    """A dictionary mapping variables to values."""
+
+    wire_map: dict = field(default_factory=dict)
+    """A map from wire values to ``AbstractQbit`` instances.
+    
+    If a value is not present in this dictionary, it needs to be extracted
+    from the ``qreg`` property.
+    """
+
+    op_math_cache: dict = field(default_factory=dict)
+    """A cache of operations that will be consumed by later operations.
+    This is a map from the ``AbstractOperator`` variables to the corresponding
+    equation. The equation will need to be interpreted when the abstract
+    operator is consumed.
+    """
+
+    def read(self, var):
```

Comment:
I'm thinking of defining a base class for jaxpr interpreters and reusing the structure for a variety of different algorithms using the template design pattern.  

You can see a prototype here:

https://github.com/PennyLaneAI/pennylane/blob/plxpr-interpreter/pennylane/capture/interpreters.py

The design and implementation is going to take a little bit of work, but in the end, we will be able to use that design to make this code much nicer :) I do see this code getting restructured once that happens.

Right now we free standing functions with a mutable input, but we could also make them all class methods. I'm fine with promoting everything to class methods now too.

Reply:
Rewriting te code to follow the structure I'm thinking of for a "Plxpr interpreter template".

Reply:
I really like the prototyped interpreters, and I think they make a great case for this design decision.
The current state of this PR left me a bit confused at first, though, even though I did review it already in an earlier state. I think this is because of the mixture of functional and class-based methods :upside_down_face: I don't mind too much, in particular if this will become more coherent with, say PL device interpreters :)

Reply:
> I think this is because of the mixture of functional and class-based methods

This is what I meant @albi3ro, further up we define a global read function for use with a local environment:
https://github.com/PennyLaneAI/catalyst/blob/92653625c521710f4cdec6116d37d1bc65996f67/frontend/catalyst/from_plxpr.py#L69-L70
https://github.com/PennyLaneAI/catalyst/blob/92653625c521710f4cdec6116d37d1bc65996f67/frontend/catalyst/from_plxpr.py#L167C9-L167C65
whereas this class duplicates both.

I do like the class though so I was hoping one could just use it for the entire interpretation procedure and get rid of the global functions. But it's not a PR blocker.